tests/providers/conftest.py

```diff
@@ -0,0 +1,89 @@
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+from __future__ import annotations
+
+from functools import lru_cache
+from pathlib import Path
+
+import pytest
+
+from tests.test_utils import db
+
+_CLEAR_DB_PROVIDERS = set()
+
+
+@lru_cache(maxsize=None)
+def providers_packages():
+    """Get providers packages full qualname."""
+
+    current_dir = Path(__file__).absolute().parent
+    providers = set()
+    for root in current_dir.iterdir():
+        if not root.is_dir():
+            continue
+
+        providers_dirs = set()
+        for sentinel in {"hooks", "operators", "sensors"}:
+            providers_dirs = providers_dirs.union({p.parent for p in root.rglob(sentinel) if p.is_dir()})
+
+        if providers_dirs:
+            for d in providers_dirs:
+                providers.add(".".join(d.relative_to(current_dir).parts))
+        else:
+            providers.add(root.name)
+
+    return providers
+
+
+def get_test_provider_name(m):
+    """Extract provider name from module full qualname."""
+    _, _, name = m.__name__.partition("providers.")
+    for provider in providers_packages():
+        if name.startswith(provider):
+            return provider
+    return None
+
+
+@pytest.fixture(scope="module", autouse=True)
+def _clear_db_between_providers_tests(request):
+    """Clear DB between each separate provider package test runs."""
+    provider_name = get_test_provider_name(request.module)
+    if provider_name and provider_name not in _CLEAR_DB_PROVIDERS:
+        _CLEAR_DB_PROVIDERS.add(provider_name)
+        db.clear_db_runs()
+        db.clear_db_datasets()
+        db.clear_db_dags()
+        db.clear_db_serialized_dags()
+        db.clear_db_sla_miss()
+        db.clear_db_pools()
+        db.clear_db_connections()
+        db.clear_db_variables()
+        db.clear_db_dag_code()
+        db.clear_db_callbacks()
+        db.clear_rendered_ti_fields()
+        db.clear_db_import_errors()
+        db.clear_db_dag_warnings()
+        db.clear_db_xcom()
+        db.clear_db_logs()
+        db.clear_db_jobs()
+        db.clear_db_task_fail()
+        db.clear_db_task_reschedule()
+        db.clear_dag_specific_permissions()
+        db.create_default_connections()
+        db.set_default_pool_slots(128)
+    yield
```

Comment:
Would it be a good idea to also clean after the provider tests? I feel like it’s generally a good principle for one thing to clean up one’s own stuffs.

Reply:
I tihnk we've came to proposal that with such (low) overhead we have we can clean up database not before/after every provider but before (after?) any module. As you can see from the attached videos above - overall overhead for running single test is neglectible if we do (~0.2 s is almost not noticeable). It will be far less if we run the whole group of tests.

If we always clean before each module - this will work not only for providers, but also for regular airflow tests - with such a low overhead, we do not need to do cleanup because we have 100% guarantee, that the DB will be cleaned before runnin any test.

Reply:
We still will need to see how much it will impact the overall test time, but I am cautiously optimistic it won't be a lot. And we can move it up to package as well if we find module is too much.

And this approach has really some interesting properties if we do such cleanup unconditionally before every module:

1) We can stop worrying about any DB side-effects outside of the modules. We have all but guarantee that any side-effects will not cross the module boundaries

2) Running all tests in a module is usually what people do - because it is easy. And whenever we see a suspicious side-effect, it will be super-easy to reproduce - just run the whole module - it will run identically in CI and locally because each module is isolated.

3) When a new user checks out and run tests, they will not have to think about initializing the database - similarly when the database has changed. So far we had `--with-db-init` switch to let airflow reset the unit test DB but it was pretty brittle and not many people even knew about its existence. We will be able to remove that now - because every time you run a test you have guarantee to get a fresh DB with the current structure.

Reply:
Another con of cleaning - this is rarely used, but not cleaning the DB after the test gives us the luxury that we can easily inspect the DB after the test completes - so for running individual tests, I think leaving the DB intact after the test has some nice debugging property as well.

Reply:
I've better keep `--with-db-init` time to time devs still need to clear DB (something changed in structure).

I have for this run configurations, but someone could find `--with-db-init` more useful rather this

![image](https://user-images.githubusercontent.com/3998685/209937487-8682130a-d1f9-474d-a5ba-5da8ce40b831.png)

Reply:
Yes. We should leave it in.